Tests/KnitTests/MainActorTests.swift

```diff
@@ -0,0 +1,180 @@
+//
+// Copyright © Block, Inc. All rights reserved.
+//
+
+import Combine
+import Knit
+import XCTest
+
+private actor ActorA { 
+    func sayHello() -> String {
+        "Hello"
+    }
+}
+
+/// A class confined to `@MainActor`
+@MainActor
+private class MainClassA { }
+
+/// Declare a custom global actor, used below
+@globalActor
+private actor CustomGlobalActor: GlobalActor {
+    
+    static var shared = CustomGlobalActor()
+
+    typealias ActorType = CustomGlobalActor
+
+}
+
+/// A class confined to a custom global actor. Means it must not be instantiated on the main thread.
+@CustomGlobalActor
+private class CustomGlobalActorClass {
+
+    /// This initializer is confined to `@CustomGlobalActor` but has a dep that is `@MainActor` confined.
+    init(mainClassA: MainClassA) {}
+
+    func sayHello() -> String {
+        "Hello"
+    }
+
+}
+
+/// A class that is async init but otherwise has sync methods.
+private class AsyncInitClass {
+
+    init() async {}
+
+    func sayHello() -> String {
+        "Hello"
+    }
+
+}
+
+/// Consumes the above types
+private class FinalConsumer {
+
+    let actorA: ActorA
+
+    let mainClassA: MainClassA
+
+    /// The dependency here is on a future of CustomGlobalActorClass, not CustomGlobalActorClass itself
+    let customGlobalActorClass: Future<CustomGlobalActorClass, Never>
+
+    var asyncInitClass: AsyncInitClass?
+
+    private var cancellables = [AnyCancellable]()
+
+    init(
+        actorA: ActorA,
+        mainClassA: MainClassA,
+        customGlobalActorClass: Future<CustomGlobalActorClass, Never>,
+        asyncInitClass: Future<AsyncInitClass, Never>
+    ) {
+        self.actorA = actorA
+        self.mainClassA = mainClassA
+        self.customGlobalActorClass = customGlobalActorClass
+
+        asyncInitClass.sink { [weak self] result in
+            self?.asyncInitClass = result
+            // Can also inform other methods that this property is now available
+        }.store(in: &cancellables)
+    }
+
+    /// Needs to be an async function due to `@CustomGlobalActor` confinement
+    func askCustomGlobalActorClassToSayHello() async -> String {
+        await customGlobalActorClass.value.sayHello()
+    }
+
+    func askActorAToSayHello() async -> String {
+        await actorA.sayHello()
+    }
+
+}
+
+private final class TestAssembly: AutoInitModuleAssembly {
+
+    typealias TargetResolver = TestResolver
+
+    func assemble(container: Container<TargetResolver>) {
+
+        container.register(
+            ActorA.self,
+            mainActorFactory: { @MainActor resolver in
+                ActorA()
+            }
+        )
+
+        container.register(
+            MainClassA.self,
+            mainActorFactory: { @MainActor resolver in
+                MainClassA()
+            }
+        )
+
+        container.register(
+            Future<CustomGlobalActorClass, Never>.self,
```

Comment:
You might want to add Future to the list of special naming cases

Reply:
Done